### PR TITLE
feat(trans): allow interpolation in tOptions if provided

### PR DESCRIPTION
**Pull request by tigerabrodi**
#### Checklist

- [x] only relevant code is changed (make a diff before you submit the PR)
- [x] run tests `npm run test`
- [ ] tests are included
- [ ] documentation is changed or added

### CODE REVIEW

**Comment by tigerabrodi** [comment]
@jamuhl As written this morning, I would work on this after work :smiley:, I am really excited :partying_face:

**Comment by tigerabrodi** [comment]
@jamuhl I have no clue why it is messing with my PR here https://codeclimate.com/github/i18next/react-i18next/pull/1204 :thinking: , if anyone else has any clues, would love some assistance :raised_hands:

**Comment by tigerabrodi** [comment]
I will make sure to update the documentation and make a PR in https://github.com/i18next/react-i18next-gitbook which will modify page https://github.com/i18next/react-i18next-gitbook/blob/master/latest/trans-component.md :tada:

File: src/Trans.js [hunk 1]
```diff
@@ -298,6 +298,7 @@ export function Trans({
     ...interpolationOverride,
     defaultValue,
     ns: namespaces,
+    ...tOptions?.interpolation,
```
Comment (JacobMGEvans) [inline src/Trans.js:301]: Would this need a fallback with the optional chaining or the `undefined` or `null` can be passed in safely?

Comment (JacobMGEvans) [reply src/Trans.js:301]: If it does, maybe `...tOptions?.interpolation ?? {} (or whatever works here)`

Comment (kentcdodds) [reply src/Trans.js:301]: Nope:

```
var x = {...null} // results in {}
var x = {...undefined} // results in {}
```

image

Comment (tigerabrodi) [reply src/Trans.js:301]: Thanks to both of you for having a look and helping me out :hugs: :pray: :gift_heart:

Comment (adrai) [reply src/Trans.js:301]: seems codeclimate has a problem with optional chaining...
alternatively, don't use the optional chaining operator and default tOptions to {}:
image

Comment (adrai) [reply src/Trans.js:301]: Can you also add a test?

**Review by adrai — CHANGES_REQUESTED** [review]

File: src/Trans.js (hunk 1 above)
Comment (tigerabrodi) [reply src/Trans.js:301]: @adrai Yeah sure :100:, I would add a test where interpolation is passed to Trans, and assert the tOptions.interpolation's value :tada:.

[the remark this answers is not in the document]
Comment (tigerabrodi) [reply src/Trans.js:301]: Thanks, @jamuhl and @adrai, will make sure to do this change and also add a test, where `tOptions.interpolation` is passed to `Trans` :raised_hands:.

Currently working, will do it after work, many thanks again, it means the world to me, my very first OSS contribution :smiley: :smiling_face_with_three_hearts:.

**Comment by tigerabrodi** [comment]
@jamuhl @adrai I did the changes as requested :100:, one issue though, it still seems not to be escaping HTML tags properly. Here is a screenshot when using `console.log(wrapper.debug())`: 
Screenshot from 2020-12-01 17-25-09


I wonder if there is a way to `console.log` and see the value of `tOptions` passed into the `Trans` component :thinking:?

Also what do you think of the changes so far? What should I do next?

Many thanks :pray: :two_hearts:

**Comment by adrai** [comment]
@tigerabrodi maybe you missed my first comment about optional chaining (`?.`): https://github.com/i18next/react-i18next/pull/1204#discussion_r533116895

regarding the test: I suspect react is already escaping your `'<no>Hello'` and not the interpolation values of i18next...

**Comment by tigerabrodi** [comment]
@adrai I actually implemented the change @jamuhl suggested https://github.com/i18next/react-i18next/pull/1204#discussion_r533110813, should I try again, but without optional chaining?

What would the next steps be, since this seems not to be working?
Is there a way to force interpolation to be prioritized over React's escaping HTML tags?

**Comment by tigerabrodi** [comment]
@adrai I pushed a new commit without Optional chaining :clap:

**Comment by adrai** [comment]
removal of optional chaining is ok, thx

but the unescape issue is a react problem 🤷‍♂️

I'm not a react expert, maybe someone knows more?

**Comment by JacobMGEvans** [comment]
It could possibly be a test environment issue. 
Possibilities:
- JSDOM isn't handling it correctly
- Whatever the test utility library being used isn't rendering it properly
- Misconfigured or missing configs for React in Jest 

@kentcdodds Any other thoughts on this? https://github.com/i18next/react-i18next/pull/1204#issuecomment-736666205

File: test/trans.render.spec.js [hunk 2]
```diff
@@ -408,3 +408,18 @@ describe('trans should not break on invalid node from translations - part2', ()
     expect(wrapper.contains(<div>&lt;hello&gt;</div>)).toBe(true);
   });
 });
+
+describe('trans getting passed tOptions.interpolation should override tOptions.interpolation', () => {
+  const stringWithHTMLTags = '<no>Hello';
+  const TestElement = () => (
+    <Trans tOptions={{ interpolation: { escapeValue: true } }}>
+      <div> {stringWithHTMLTags} </div>
```
Comment (kentcdodds) [inline test/trans.render.spec.js:416]:
```suggestion
      <div>{stringWithHTMLTags}</div>
```

Maybe the extra spaces are the problem?

**Review by kentcdodds — COMMENTED** [review]
I'd suggest switching to React Testing Library eventually, but maybe this is the issue?

File: test/trans.render.spec.js (hunk 2 above)
Comment (JacobMGEvans) [reply test/trans.render.spec.js:416]: Oh, I didn't even notice the spaces! 😅

**Comment by adrai** [comment]
> * Whatever the test utility library being used isn't rendering it properly
> I'd suggest switching to React Testing Library eventually, but maybe this is the issue?

probably it's an [enzyme](https://github.com/enzymejs/enzyme) issue...

@tigerabrodi if it works for you in react, I think if this can't be fixed with enzyme, we can skip the test or if you like you can try to switch to a better react test library?

File: test/trans.render.spec.js (hunk 2 above)
Comment (tigerabrodi) [reply test/trans.render.spec.js:416]: I would try this first, thanks both of you :two_hearts:

**Comment by tigerabrodi** [comment]
> > * Whatever the test utility library being used isn't rendering it properly
> >   I'd suggest switching to React Testing Library eventually, but maybe this is the issue?
> 
> probably it's an [enzyme](https://github.com/enzymejs/enzyme) issue...
> 
> @tigerabrodi if it works for you in react, I think if this can't be fixed with enzyme, we can skip the test or if you like you can try to switch to a better react test library?

@adrai You mean for that particular test use React Testing Library? Or convert the whole project to use React Testing Library, which I would actually be down to :smiley: ?

My suggestion: Try out what @kentcdodds mentioned, if it still does not work, skip the test and I would with joy and passion convert this project to use React Testing Library instead of Enzyme :+1:.

Side question: @adrai I was wondering if this project plans on migrating fully to TypeScript? If so, I'd also be down to help in every way I can :tada:. We use it at work by the way, and you and @jamuhl are just really awesome, being able to contribute here just means the world to me :smiley: :two_hearts:.

**Comment by adrai** [comment]
@tigerabrodiI would say, if you really like to change all tests to "React Testing Library" try it 😉

about TypeScript, idk, maybe not, need to wait for @jamuhl answer tomorrow morning.

**Comment by tigerabrodi** [comment]
> @tigerabrodiI would say, if you really like to change all tests to "React Testing Library" try it 
> 
> about TypeScript, idk, maybe not, need to wait for @jamuhl answer tomorrow morning.

@adrai Is it okay to merge the changes I made in `Trans` component first and make a new PR with the conversion over to React Testing Library?

I would make a fixup commit and squash the commits, so we only have the changes in `Trans` component and nothing else :+1:.

In a separate PR, I would start the conversion to React Testing Library :tada:. I plan on doing it with @JacobMGEvans, we really look forward to it :two_hearts: :star_struck:

**Comment by marcosvega91** [comment]
Good job @tigerabrodi :)

**Comment by tigerabrodi** [comment]
@adrai I pushed the last commit :raised_hands:

Do I still need to change anything or?

In another PR, I would start the conversion to React Testing Library :tada: :two_hearts:

**Review by adrai — APPROVED** [review]

**Comment by adrai** [comment]
lgtm

**Comment by jamuhl** [comment]
Looks awesome 👍... a great first contribution...keep it up...we're all happy to get help.

Regarding the encoding. This comes from react which does a good job of guarding people against accidentally pass unencoded user input into the page (XSS attacks):

image

So it comes to two options a dev has to insert translations into a page:

a) using dangerouslysetinnerhtml https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml
b) using the Trans component (https://react.i18next.com/latest/trans-component) with it's `<Trans><div><no />hello</div></Trans> to string "<0><0 />hello</0>"` conversion and interpolation of react children's or named replacement like in `<Trans defaults="<no/> hello"  components={{ no: <no /> }} />` 

A little hard to simply describe how the Trans component works / can be used -> which might be a cause of some pain: https://github.com/i18next/react-i18next-gitbook/issues/15

So while this PR fixes an issue - I guess, it will not solve the initial problem you liked to solve. You might open an issue with the content you like to get translated and we might help out with some options on how to use the Trans component for it.

Regarding the switch to `React Testing Library` that would be the best Xmas present contribution to us, I could think of. If you like to do so - go for it.

Switching completely to typescript is something we considered - @adrai started work on a successor for the current i18next implementation - first in typescript but on the way, we decided to no go further with typescript as it felt more like a burden than a win (not sure it's a lack of our understanding of typescript or really just the price you pay for types, idk).

**Comment by jamuhl** [comment]
Thank you for your contribution - looking forward to see more from you - well done.

This fix was published in react-i18next@11.7.4

**Comment by tigerabrodi** [comment]
@jamuhl hahaha, wohooo :champagne: :tada:, you are awesome, I will never forget this moment nor you :raised_hands: :two_hearts:

**Comment by tigerabrodi** [comment]
@jamuhl I think regarding TypeScript, that would not be necessary, but React Testing Library, wohooo :tada:, that'd be really awesome, my friends are also down to help me out @kentcdodds @JacobMGEvans and many more :two_hearts:

Thank you again @jamuhl, really, my first step into OSS, I admire it so much, I look up to so many, @kentcdodds himself is one of my greatest inspiration and role model, this really means the world to me, I look forward to my long journey here on github haha :smiley: :100: :tada: :champagne:

**Comment by tigerabrodi** [comment]
@jamuhl Regarding the issue we have at work, I would open an issue for that and try to explain it as thoroughly as possible, thank you again, and thank you @adrai also, you guys have no idea how much this means to me!!! :two_hearts: Regarding converting to React Testing Library, that is the least I can do, someday we will all meet up in a React conf hopefully hahaha. This really means so much to me, more than every single birthday that I have had and will have :smiley: :tada:

**Comment by tigerabrodi** [comment]
@jamuhl Regarding the conversion over to TS, friends from @kentcdodds Discord also wanna help out :partying_face:. I plan on opening the PR, and from there magic will happen :smiley: :100:, I look forward to it, coding with friends is always awesome :smile:.

Thanks to you and @adrai, my first big step into OSS :two_hearts:

**Pull request merged**